### PR TITLE
Allow configuring ICMP redirects on a Vlan Interface

### PR DESCRIPTION
**Pull request by mat128**

### CODE REVIEW

**Comment by mat128** [comment]
I have mixed feelings about the TODOs in switch_base but felt this was the right place for this. I guess I could alternatively create an issue on github for the nomenclature violations. What do you guys think?

File: netman/adapters/switches/cached.py [hunk 1]
```diff
@@ -338,3 +338,10 @@ def remove_dhcp_relay_server(self, vlan_number, ip_address):
 
     def enable_lldp(self, interface_id, enabled):
         self.real_switch.enable_lldp(interface_id, enabled)
+
+    def set_vlan_icmp_redirects_state(self, vlan_number, state):
+        self.real_switch.set_vlan_icmp_redirects_state(vlan_number, state)
+        try:
+            self.vlans_cache[vlan_number].icmp_redirects = state
+        except ValueError:
```
Comment (lindycoder) [inline netman/adapters/switches/cached.py]: This try except is not needed

File: netman/adapters/switches/remote.py [hunk 2]
```diff
@@ -257,7 +257,14 @@ def remove_dhcp_relay_server(self, vlan_number, ip_address):
...
     def enable_lldp(self, interface_id, enabled):
         self.put("/interfaces/{}/lldp".format(interface_id),
-                 raw_data={True: "true", False: "false"}[enabled])
+                 raw_data=self.get_json_boolean(enabled))
+
+    def set_vlan_icmp_redirects_state(self, vlan_number, state):
+        self.put('/vlans/{}/icmp-redirects'.format(vlan_number),
+                 raw_data=self.get_json_boolean(state))
+
+    def get_json_boolean(self, state):
```
Comment (lindycoder) [inline netman/adapters/switches/remote.py]: This can be outside the class and i suggest prefixing with _

**Comment by lindycoder** [comment]
About the TODOs, every called not covered by the compliance test is "to do" and implementing it should force us to format the method name accordingly.   That being said, we could have an issue as you said,  make one pass adding the new method names and deprecating the other right now for later.

I think raising the issue can't hurt

**Comment by lindycoder** [comment]
But they're definately not part of this PR i'd appreciate if they were removed

File: netman/core/objects/vlan.py [hunk 3]
```diff
@@ -26,3 +26,7 @@ def __init__(self, number=None, name=None, ips=None, vrrp_groups=None, vrf_forwa
         self.ips = ips or []
         self.vrrp_groups = vrrp_groups or []
         self.dhcp_relay_servers = dhcp_relay_servers or []
+        if icmp_redirects is not None:
```
Comment (lindycoder) [inline netman/core/objects/vlan.py]: This is an interesting point.  The hardware we know have an icmp redirects state defaulting to True.  But that might not always be the case...

The adapter which instantiate the vlan is responsible to know that, but it should now leave this empty, even if instantiated empty for ease of code.

Therefore i recommend not having a default value here, let the adapter set a value, the compliance test will ensure it does

File: tests/adapters/compliance_tests/set_vlan_icmp_redirects_state_test.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+from tests.adapters.configured_test_case import ConfiguredTestCase
+
+
+class SetVlanIcmpRedirectsStateTest(ConfiguredTestCase):
+    _dev_sample = "cisco"
+
+    def setUp(self):
+        super(SetVlanIcmpRedirectsStateTest, self).setUp()
+        self.client.add_vlan(1000)
+        self.client.add_vlan(1001)
```
Comment (lindycoder) [inline tests/adapters/compliance_tests/set_vlan_icmp_redirects_state_test.py]: no reason to use 2 vlans here

File: tests/adapters/compliance_tests/set_vlan_icmp_redirects_state_test.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        vlan = self.get_vlan_from_list(1000)
+        assert_that(vlan.number, is_(1000))
+        assert_that(vlan.icmp_redirects, is_(True))
+
+    def test_disables_icmp_redirects_when_given_false(self):
+        self.client.set_vlan_icmp_redirects_state(1001, False)
+        vlan = self.get_vlan_from_list(1001)
+        assert_that(vlan.number, is_(1001))
+        assert_that(vlan.icmp_redirects, is_(False))
+
```
Comment (lindycoder) [inline tests/adapters/compliance_tests/set_vlan_icmp_redirects_state_test.py]: Missing tests

Vlan does not exist

File: tests/adapters/switches/cisco_test.py [hunk 6]
```diff
@@ -2450,3 +2450,33 @@ def test_remove_non_existent_dhcp_relay_server_fails(self):
...
+            "Enter configuration commands, one per line.  End with CNTL/Z."
+        ])
+        self.mocked_ssh_client.should_receive("do").with_args("interface vlan 1234").and_return([]).once().ordered()
+        self.mocked_ssh_client.should_receive("do").with_args("no shutdown").and_return([]).once().ordered()
+        self.mocked_ssh_client.should_receive("do").with_args("no ip redirects").and_return([]).once().ordered()
+        self.mocked_ssh_client.should_receive("do").with_args("exit").and_return([]).twice().ordered().ordered()
+        self.mocked_ssh_client.should_receive("do").with_args("write memory").and_return([]).once().ordered()
+
+        self.switch.set_vlan_icmp_redirects_state(1234, False)
+
```
Comment (lindycoder) [inline tests/adapters/switches/cisco_test.py]: Missing tests:

Vlan does not exist
Interface Vlan does not exist and need to be created

File: netman/adapters/switches/cisco.py [hunk 7]
```diff
@@ -472,6 +479,9 @@ def apply_interface_running_config_data(vlan, data):
         elif regex.match("^ ip helper-address ([^\s]*)", line):
             vlan.dhcp_relay_servers.append(IPAddress(regex[0]))
 
+        elif regex.match("^ no ip redirects", line):
+            vlan.icmp_redirects = False
```
Comment (lindycoder) [inline netman/adapters/switches/cisco.py]: This line is not unit tested

**Comment by lindycoder** [comment]
There's a new attribute to the vlan,  you should add it to the get_vlan compliance test

File: netman/adapters/switches/cached.py (hunk 1 above)
Comment (mat128) [inline netman/adapters/switches/cached.py]: Fixed in revision 2

File: netman/adapters/switches/cisco.py (hunk 7 above)
Comment (mat128) [inline netman/adapters/switches/cisco.py]: Covered in revision 2 by updating get_vlan tests

File: netman/adapters/switches/remote.py (hunk 2 above)
Comment (mat128) [inline netman/adapters/switches/remote.py]: Fixed in revision 2

File: netman/core/objects/vlan.py (hunk 3 above)
Comment (mat128) [inline netman/core/objects/vlan.py]: Done in revision 2.

File: tests/adapters/compliance_tests/set_vlan_icmp_redirects_state_test.py (hunk 4 above)
Comment (mat128) [inline tests/adapters/compliance_tests/set_vlan_icmp_redirects_state_test.py]: Fixed.

File: NOMENCLATURE.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+Name should contain parent resource (e.g. interface_port_mode, interface_access_vlan)
+
+### Setting
+* Should be called set_*_state
+* Should overwrite if already set
+
+### Unsetting
+* Should be called unset_*
+* Should not raise if not set
+
```
Comment (stephanerobert) [inline NOMENCLATURE.md]: Is the unsetting really useful? Looking below, I see that we pass the state along side the set
ie: set_vlan_icmp_redirects_state(vlan_number, state) where state can be True or False. So what would be the use, in this Boolean context, of an unsetting method?

File: tests/adapters/switches/cisco_test.py (hunk 6 above)
Comment (mat128) [inline tests/adapters/switches/cisco_test.py]: Fixed in revision 2.

File: tests/adapters/compliance_tests/set_vlan_icmp_redirects_state_test.py (hunk 5 above)
Comment (mat128) [inline tests/adapters/compliance_tests/set_vlan_icmp_redirects_state_test.py]: Fixed.

File: NOMENCLATURE.md (hunk 8 above)
Comment (lindycoder) [inline NOMENCLATURE.md]: To reset to vendor's default behavior

Comment (mat128) [inline NOMENCLATURE.md]: We also have a bunch of "remove_" where what they really do is unset. See issue #52

File: netman/core/objects/vlan.py [hunk 9]
```diff
@@ -26,3 +26,4 @@ def __init__(self, number=None, name=None, ips=None, vrrp_groups=None, vrf_forwa
         self.ips = ips or []
         self.vrrp_groups = vrrp_groups or []
         self.dhcp_relay_servers = dhcp_relay_servers or []
+        self.icmp_redirects = icmp_redirects
```
Comment (lindycoder) [inline netman/core/objects/vlan.py]: EOF new line, set this in your editor :)

**Comment by lindycoder** [comment]
:+1:

File: netman/core/objects/vlan.py (hunk 9 above)
Comment (mat128) [inline netman/core/objects/vlan.py]: fixed (--amend of last commit)

**Comment by mat128** [comment]
Appreciate the good feedback :) @stephanerobert what do you think?

File: tests/adapters/switches/cisco_test.py [hunk 10]
```diff
@@ -2450,3 +2454,89 @@ def test_remove_non_existent_dhcp_relay_server_fails(self):
...
+        self.command_setup()
+
+        self.mocked_ssh_client.should_receive("do").with_args("show running-config interface vlan 1234").once().ordered().and_return([
+            "Building configuration...",
+            "Current configuration : 41 bytes",
+            "!",
+            "interface Vlan1234",
+            " no ip address",
+            "end"
+        ])
```
Comment (stephanerobert) [inline tests/adapters/switches/cisco_test.py]: Shouldn't we see a "no ip redirects" in here?

**Comment by lindycoder** [comment]
:+1:

**Comment by stephanerobert** [comment]
:+1:

**Pull request merged**